File: src/ovi_voice_assistant/agent/assistant_context.py
"""Run context passed to agent tools for device interaction."""

from __future__ import annotations

import asyncio
import logging
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ovi_voice_assistant.memory import Memory
    from ovi_voice_assistant.music import MusicGroup, MusicPlayer
    from ovi_voice_assistant.scheduler import Scheduler

logger = logging.getLogger(__name__)


@dataclass
class AssistantContext:
    """Context available to all agent tools during a pipeline run.

    Created by ``VoiceAssistant.create_context`` — one per device connection.
    The ``announce`` callback drives a full voice assistant event sequence
    so the device activates its speaker and plays the audio.
    """

    announce: Callable[[str], asyncio.Task] | None = None
    """Trigger a TTS announcement on the device. Returns the background task."""

    say: Callable[[str], Awaitable[None]] | None = None
    """Speak text immediately during a pipeline run without interrupting it."""

    music_player: MusicPlayer | None = None
    """Music player for this device — controls queue, playback state."""

    music_group: MusicGroup | None = None
    """Shared music group for synchronized multi-device playback."""

    scheduler: Scheduler | None = None
    """Scheduler for creating/managing proactive automations."""

    memory: Memory | None = None
    """Persistent memory for fact extraction and recall across conversations."""

    _timers: dict[str, tuple[asyncio.TimerHandle, float]] = field(default_factory=dict)
    """Maps label → (handle, fire_time) where fire_time is loop.time() epoch."""

    _pending_say: asyncio.Task | None = field(default=None, init=False, repr=False)
    """Background task for in-flight say() audio."""

    async def drain_say(self) -> None:
        """Wait for any in-flight say() audio to finish."""
        task = self._pending_say
        self._pending_say = None
        if task is not None and not task.done():
            try:
                await task
            except Exception:
                logger.debug("say() task failed", exc_info=True)

    def schedule_timer(self, seconds: float, label: str) -> None:
        """Schedule a timer that announces on the device when it fires."""
        loop = asyncio.get_running_loop()
        handle = loop.call_later(
            seconds,
            lambda: loop.create_task(self._on_timer_fire(label)),
        )
        self._timers[label] = (handle, loop.time() + seconds)
        logger.info("Timer scheduled: %s (%.0fs)", label, seconds)

    def cancel_timer(self, label: str) -> bool:
        """Cancel a named timer. Returns True if found."""
        entry = self._timers.pop(label, None)
        if entry:
            entry[0].cancel()
            logger.info("Timer cancelled: %s", label)
            return True
        return False

    def get_timer_status(self) -> dict[str, float]:
        """Return remaining seconds for each active timer."""
        now = asyncio.get_running_loop().time()
        return {
            label: max(0.0, fire_time - now)
            for label, (_, fire_time) in self._timers.items()
        }

    async def _on_timer_fire(self, label: str) -> None:
        """Announce timer completion on the device."""
        self._timers.pop(label, None)
        text = f"Your {label} timer is done."
        logger.info("Timer fired: %s", label)
        if self.announce:
            self.announce(text)
        else:
            logger.warning("No announce callback — cannot notify device")

File: internal/agent/assistant_context.go
// Package agent implements the conversational agent: an OpenAI-compatible
// tool-calling loop with built-in tools, MCP servers and sub-agents.
package agent

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/bryfur/ovi-voice-assistant/internal/memory"
	"github.com/bryfur/ovi-voice-assistant/internal/music"
	"github.com/bryfur/ovi-voice-assistant/internal/scheduler"
)

// MemoryStore is the memory surface used by the agent.
type MemoryStore interface {
	Recall(ctx context.Context, query string, budget memory.Budget, maxTokens int) (memory.RecallResult, error)
	Retain(ctx context.Context, content, extraContext string) (memory.RetainResult, error)
}

// Context is the run context available to all agent tools during a
// pipeline run. One is created per device connection.
type Context struct {
	// Announce triggers a TTS announcement on the device (fire-and-forget).
	Announce func(text string)
	// Say speaks text immediately during a pipeline run without interrupting it.
	Say func(ctx context.Context, text string) error
	// MusicPlayer is the per-device music player.
	MusicPlayer *music.MusicPlayer
	// MusicGroup is the shared group for synchronized multi-device playback.
	MusicGroup *music.MusicGroup
	// Scheduler creates/manages proactive automations.
	Scheduler *scheduler.Scheduler
	// Memory is persistent memory for fact extraction and recall.
	Memory MemoryStore

	mu     sync.Mutex
	timers map[string]*timerEntry
}

type timerEntry struct {
	timer    *time.Timer
	fireTime time.Time
}

// ScheduleTimer schedules a timer that announces on the device when it fires.
func (c *Context) ScheduleTimer(seconds float64, label string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.timers == nil {
		c.timers = map[string]*timerEntry{}
	}
	if old, ok := c.timers[label]; ok {
		old.timer.Stop()
	}
	d := time.Duration(seconds * float64(time.Second))
	entry := &timerEntry{fireTime: time.Now().Add(d)}
	entry.timer = time.AfterFunc(d, func() { c.onTimerFire(label) })
	c.timers[label] = entry
	slog.Info("Timer scheduled", "label", label, "seconds", int(seconds))
}

// CancelTimer cancels a named timer. Returns true if found.
func (c *Context) CancelTimer(label string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.timers[label]
	if !ok {
		return false
	}
	entry.timer.Stop()
	delete(c.timers, label)
	slog.Info("Timer cancelled", "label", label)
	return true
}

// TimerStatus returns remaining seconds for each active timer.
func (c *Context) TimerStatus() map[string]float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	now := time.Now()
	out := make(map[string]float64, len(c.timers))
	for label, e := range c.timers {
		remaining := e.fireTime.Sub(now).Seconds()
		if remaining < 0 {
			remaining = 0
		}
		out[label] = remaining
	}
	return out
}

// onTimerFire announces timer completion on the device.
func (c *Context) onTimerFire(label string) {
	c.mu.Lock()
	delete(c.timers, label)
	announce := c.Announce
	c.mu.Unlock()
	text := fmt.Sprintf("Your %s timer is done.", label)
	slog.Info("Timer fired", "label", label)
	if announce != nil {
		announce(text)
	} else {
		slog.Warn("No announce callback — cannot notify device")
	}
}
